content/blog/2019-09-25-vril-vector-bril.md

```diff
@@ -0,0 +1,139 @@
+
+In order to move data from an element of an array `arr[idx]` to a variable `var`, we can use `a2v` op:
+```
+var: int = a2v arr idx
+```
+Similarly, we can move a value stored in a variable `var` into one of the array's element `arr[idx]`:
+```
+arr: int v2a var index
+```
+
+To put a constant into an array, the constant should first be put to a variable.
+To operate with a scalar operator (for instance `print`) on a single element, the array element has to first be put to a variable.
+
+Vector ops
+----------------------------
+
+### Vector ops in Vril
+There are two types of vector ops in Vril: configuration ops and arithmetic/logic ops. Vector arithmetic or logic operations take arrays as arguments and produce an array or a scalar value as a result. Instead, configuration ops help in configuring a vector state.
+
+### Adding vector ops to Vril
+For now there is only one configuration op: `setvl`. 
+```
+vl: int setvl val
+```
+is used to request `val` lanes to the IR and `setvl` returns `vl:=val` if `vl<maxvl` or `vl:=maxvl` if `vl>=maxvl`.
+We have also implemented only one arithmetic vector op `vadd` as a proof of concept:
+```
+arr3: array vadd arr1 arr2
+```
+Which takes two arrays as arguments and performs element-wise vector addition, and stores it into a third array.
+
+### Simulating vector ops in Vrili (Vril interpreter)
+We have extended Brili to interpret vector operations.
+Vrili executes for loops across the different elements of the array operands in order to produce the correct result. This emulates what a vector microarchitecture would do concurrently with a single fetch-decode effort.
+
+Evaluation
+----------------------------
+
+Vril makes two major additions to Bril, adding Arrays and extending Briloperations with vector operations. This section describes how we evaluate each of these.
+
+### Evaluating Arrays
+We take three approaches to evaluate arrays,
+- evaluating integration to Bril
+- qualitative evaluation to justify arrays
+- evaluating functionality of arrays
+
+#### Integration tests for arrays
+We ran tests to generate JSON representation of Bril from text format and vice-versa to evaluate whether the array specification is expressible in both available formats. These tests reside in `vtest/parse` and `vtest/print` directories.
+
+#### Qualitative evaluation of arrays
+Arrays provide a convenient method to use loops in Bril. To add two groups of 10 numbers, conventional Brili required 44 lines of code, 20 of which would define the 20 numbers and 10 to add them. Vril required 50 lines of code, 20 of which were still used to define two arrays element by element and 6 lines to add them in a loop body. It is observed that as the size of the arrays grow, conventional Bril would need correspondingly more lines of code where as Vril would still use 6 lines of code in the loop body.
```

Comment:
FWIW, I don't think it takes much careful argument to justify the existence of arrays—the purpose and usefulness of arrays is pretty well-understood. Even so, I like the idea of showing that the number of instructions is smaller.

Reply:
Noted.

Reply:
As to why, it's trying to address https://github.com/sampsyo/cs6120/pull/45/files#r330825016. I think it's better to link this with the results table.